src/Microsoft.AspNet.Mvc.ViewFeatures/Buffer/MemoryPoolViewBufferScope.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Mvc.ViewFeatures.Buffer
+{
+    /// <summary>
+    /// A <see cref="IViewBufferScope"/> that uses pooled memory.
+    /// </summary>
+    public class MemoryPoolViewBufferScope : IViewBufferScope, IDisposable
+    {
+        public static readonly int SegmentSize = 512;
+        private readonly ArrayPool<ViewBufferValue> _pool;
+        private List<ViewBufferValue[]> _leased;
+        private bool _disposed;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="MemoryPoolViewBufferScope"/>.
+        /// </summary>
+        /// <param name="pool">The <see cref="ArrayPool{ViewBufferValue}"/> for creating
+        /// <see cref="ViewBufferValue"/> instances.</param>
+        public MemoryPoolViewBufferScope(ArrayPool<ViewBufferValue> pool)
+        {
+            _pool = pool;
+        }
+
+        /// <inheritdoc />
+        public ViewBufferValue[] GetSegment()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(typeof(MemoryPoolViewBufferScope).FullName);
+            }
+
+            if (_leased == null)
+            {
+                _leased = new List<ViewBufferValue[]>(1);
+            }
+
+            ViewBufferValue[] segment = null;
+
+            try
+            {
+                segment = _pool.Rent(SegmentSize);
```

Comment:
We probably don't need to implement chunking since the pool will go pretty low if we need it to. Default minimum is 16.